packages/gatsby/src/utils/node-descriptor.js

```diff
@@ -0,0 +1,346 @@
+/*
+## Incrementally track the structure of nodes with metadata
+
+This metadata can be later utilized for schema inference
+(via building `exampleValue` or directly)
+
+### Usage example:
+
+```javascript
+  const node1 = { id: '1', foo: 25, bar: 'str' }
+  const node2 = { id: '1', foo: 'conflict' }
+
+  let meta = { ignoredFields: new Set(['id']) }
+  meta = addNode(meta, node1)
+  meta = addNode(meta, node2)
+  console.log(meta.fieldMap)
+  // outputs: {
+  //   foo: {
+  //     int: { total: 1, example: 25 },
+  //     string: { total: 1, example: 'conflict' },
+  //   },
+  //   bar: {
+  //     string: { total: 1, example: 'str' },
+  //   },
+  // }
```

Comment:
How are nested objects represented here? How arrays are represented?

I.e. 
what would
```
const node2 = { id: '1', nested: { foo: 'bar' }, array: [1, 2, 3, "string"] }
```
produce

(if this is handled already)

Reply:
Ah there are types below - ignore above ;)

Reply:
Yeah, those are just simple usage examples, didn't want to distract with all the recursion stuff here %)